session_test.go

```diff
@@ -67,20 +77,19 @@ func (p *pipeConn) UnblockWrites() {
 	<-p.writeBlocker
 }
 
-func testConn() (io.ReadWriteCloser, io.ReadWriteCloser) {
-	read1, write1 := io.Pipe()
-	read2, write2 := io.Pipe()
-	conn1 := &pipeConn{
-		reader:       read1,
-		writer:       write2,
-		writeBlocker: make(chan struct{}, 1),
-		closeCh:      make(chan struct{}, 1),
-	}
-	conn2 := &pipeConn{
-		reader:       read2,
-		writer:       write1,
-		writeBlocker: make(chan struct{}, 1),
-		closeCh:      make(chan struct{}, 1),
+func testConn() (conn1, conn2 net.Conn) {
```

Comment:
These tests made a _lot_ of assumptions about the synchronous nature of this connection. I've had to rewrork most of them to handle the fact that all writes are now async.